### PR TITLE
pthread: add setname/getname extension functions

### PR DESCRIPTION
**Pull request by nslowell**
Adding the ability to set and get pthread names by defining
some non-portable extension functions based on the glibc
equivalent.

Signed-off-by: Nicholas Lowell <nlowell@lexmark.com>

### CODE REVIEW

**Review by pabigot — COMMENTED** [review]
Unless there's precedent for non-POSIX functions in the Zephyr POSIX API, I wouldn't take this.  But this isn't my area so that's just a comment.

**Comment by nslowell** [comment]
While I do appreciate the desire to only use standard POSIX, setname/getname has been around for nearly a decade and is implemented in a broad range of POSIX implementations.  Additionally, setting thread names is a feature provided by the zephyr kernel and it seems logical to provide a plug-in to that functionality even when using pthreads.

**Comment by pfalcon** [comment]
I see the purpose of POSIX API as two-fold: a) to provide an API sets which are on the opposite spectrum of a typical RTOS approach of NIH APIs; b) be useful to port real-world applications, and thus include established extensions and de-facto standard functionality.

In this regard, this extension would be useful. But useful for what? That's exactly the question which this PR (or more specifically, commit), doesn't answer. Adding more things just for the purpose of adding them isn't that useful. But if these new API helped to actually port some existing GLIBC software to Zephyr, then well, this PR (well, commit), should explain that, and be the basis of upvoting and upholding this addition.

**Comment by nslowell** [comment]
Hopefully the additional information in the commit message is what you're looking for.

File: include/posix/pthread.h [hunk 1]
```diff
@@ -522,6 +522,10 @@ int pthread_key_delete(pthread_key_t key);
 int pthread_setspecific(pthread_key_t key, const void *value);
 void *pthread_getspecific(pthread_key_t key);
 
+/* Extension Functions */
```
Comment (pfalcon) [inline include/posix/pthread.h]: Please clarify as "Glbc extension functions".

File: lib/posix/pthread.c [hunk 2]
```diff
@@ -597,3 +597,46 @@ int pthread_attr_destroy(pthread_attr_t *attr)
...
+ * @brief Set thread name. Non-portable, extension function
+ *
+ */
+int pthread_setname_np(pthread_t thread, const char *name)
+{
+#ifdef CONFIG_THREAD_NAME
+	k_tid_t kthread = (k_tid_t)thread;
+
+	if ((kthread == NULL)) {
+		return ESRCH;
```
Comment (pfalcon) [inline lib/posix/pthread.c]: "man pthread_setname_np" doesn't explicitly mention ESRCH as a return code. Based on the source code, Glibc would return ENOENT in this case: https://code.woboq.org/userspace/glibc/sysdeps/unix/sysv/linux/pthread_setname.c.html

**Review by pfalcon — COMMENTED** [review]
> Hopefully the additional information in the commit message is what you're looking for.

Fairly speaking, I don't see much additional info added. But I'm personally happy to +1 addition of any API which is more or less standard rather than adhoc to Zephyr. But in that regard, some extra attention to API contract (error codes, etc.) might be helpful. Please also review codestyle reports - at least some issues there are fixable.

And as ultimate way to make this great PR, I should mention usefulness of unit tests. I would point to https://github.com/zephyrproject-rtos/zephyr/pull/22863 as a recent example of introducing new POSIXish functionality together with a test, instead of leaving it "for later". Perhaps, we can invite more people to follow suit.

**Comment by nslowell** [comment]
I appreciate the feedback.

I will see what I can do to address these things.  
I hesitate to highlight Glibc too much because though they were the first, there have been various adoptions. https://stackoverflow.com/a/7989973
That's why it isn't technically a Glibc extension, but Zephyr's own extension (that just happens to closely match many other implementations).

It actually looks like Oracle was referenced for Zephyr's posix implementation (solely based on its return of ESRCH from similar functions) https://docs.oracle.com/cd/E37838_01/html/E61057/tlib-1.html#OSMPGgotxm

I will fix the unnecessary parentheses but prefer keeping ESRCH positive to be consistent with the other posix functions.

Stay tuned....

**Comment by pfalcon** [comment]
> but prefer keeping ESRCH positive to be consistent with the other posix functions.

Ack on this, pthread functions return errno's directly, and those are known to be positive.

**Comment by pfalcon** [comment]
> I hesitate to highlight Glibc too much because though they were the first, there have been various adoptions. https://stackoverflow.com/a/7989973

That's a very sad post, because it shows stuff like:

~~~
// Linux, NetBSD:
int pthread_getname_np(pthread_t th, char *buf, size_t len);
// some implementations don't have a safe buffer (see MKS/IBM below)
int pthread_getname_np(pthread_t thread, const char **name);
int pthread_getname_np(pthread_t thread, char *name);
~~~

Seeing that, let me humbly but firmly insist that it was mentioned as "Glibc extension". So that when in some time another guys comes and says "Hey, you got the prototype for pthread_getname_np() wrong", it was very clear (to anyone) that nope, we have it right, because we follow Glibc.

> That's why it isn't technically a Glibc extension, but Zephyr's own extension (that just happens to closely match many other implementations).

Formulated like that, I'm with @pabigot on this matter: we should not have Zephyr-specific POSIX extensions. Again, where my position differs is that IMHO, we can provide well-known, well-vetted extensions, as implemented in popular, well-vetted projects/systems. Glibc is clearly example of such a project, we can easily believe that if they added functions like that, there were good reasons for that, and stuff like function signatures was discussed and well-reviewed. To invoke this exception on adding non-standard, but POSIXish stuff, referring to well-known project is imperative (for me), to show it's not some random NIH.

In general, you already may have noticed that different developers have different opinions re: stuff to add to Zephyr. And I'm trying to make a review from the following PoV: what if, after this patch, some other guy submits patch which just removes these APIs. In the current situation, such a patch would be about as good as a patch to add them. I'm trying to make sure that this patch provides enough detail and information to judge attempt at removal as a clear regression of Zephyr functionality.

File: lib/posix/pthread.c [hunk 3]
```diff
@@ -597,3 +597,76 @@ int pthread_attr_destroy(pthread_attr_t *attr)
 
 	return EINVAL;
 }
+
+/**
+ * @brief Set name of POSIX thread.
```
Comment (pfalcon) [inline lib/posix/pthread.c]: Sorry for finding more things to comment on, but Zephyr has a pattern of providing docstring in .h files. I'm not even sure if doxygen will pick up docs from source files. Even if does, myself and other developers got used to look up info in .h files, so could you please move docstring there. (Don't get confused by the lack of docstrings to other POSIX(ish) funcs - we kinda assume people could find reference info somewhere else, but eventually would like to add explicit docstrings).

**Comment by pfalcon** [comment]
> It actually looks like Oracle was referenced for Zephyr's posix implementation (solely based on its return of ESRCH from similar functions) https://docs.oracle.com/cd/E37838_01/html/E61057/tlib-1.html#OSMPGgotxm

I see, so you used ESRCH because it's consistent with its usage in other existing functions in Zephyr pthread impl. Sounds good. But if someone later comes up with a patch to change that to ENOENT with argument of Glibc compatibility, I'd be positive about it.

**Comment by nslowell** [comment]
I greatly appreciate your patience, guidance, and feedback.

**Review by pfalcon — APPROVED** [review]
Looks good, thanks!

**Comment by nslowell** [comment]
If the failed Shippable is my fault somehow, please advise.

**Comment by pfalcon** [comment]
> If the failed Shippable is my fault somehow, please advise.

Nope, looks like a random timeout. The way to restart CI is to close/reopen a PR, doing that.

**Comment by nslowell** [comment]
@andrewboie Are you ok with this PR?  Some of the unfixed warnings are in order to stay consistent with the rest of the file.

**Review by tyrantlizard77 — APPROVED** [review]

**Pull request merged**